include/multipass/exceptions/formatted_exception_base.h

```diff
@@ -0,0 +1,140 @@
+        std::is_constructible<BaseExceptionType, std::string>::value ||
+            std::is_constructible<BaseExceptionType, std::error_code, std::string>::value,
+        "BaseExceptionType must either be constructible with (std::string) or (std::error_code, std::string).");
+    static_assert(std::is_base_of<std::exception, BaseExceptionType>::value,
+                  "BaseExceptionType must derive from std::exception");
+
+    /**
+     * Produces a formatted string using fmt and args.
+     *
+     * Possible format exceptions are handled gracefully.
+     *
+     * @tparam Args Format argument types
+     *
+     * @param [in] fmt Format string
+     * @param [in] args Format arguments
+     */
+    template <typename... Args>
+    FormattedExceptionBase(fmt::format_string<Args...> fmt, Args&&... args)
+        : BaseExceptionType(failsafe_format(fmt, std::forward<Args>(args)...))
+    {
+    }
+
+    /**
+     * Produces a formatted string using fmt and args.
+     *
+     * Possible format exceptions are handled gracefully.
+     *
+     * @tparam Args Format argument types
+     *
+     * @param [in] ec Error code
+     * @param [in] fmt Format string
+     * @param [in] args Format arguments
+     */
+    template <typename... Args>
+    FormattedExceptionBase(std::error_code ec, fmt::format_string<Args...> fmt, Args&&... args)
+        : BaseExceptionType(ec, failsafe_format(fmt, std::forward<Args>(args)...))
+    {
+    }
+
+private:
+    /**
+     * Exception catcher for fmt::format.
+     *
+     * Since this class is an exception class itself, it cannot simply throw
+     * exceptions on the constructor, otherwise it'll lead to `std::terminate()`
+     * and we definitely don't want to terminate just because of a recoverable
+     * format error.
+     *
+     * Instead, the code catches the exception and returns a string that indicate
+     * that there was a format error with the given format string, so it can be
+     * traced and fixed.
+     *
+     * The function is not marked as `noexcept` since in theory the std::string
+     * constructor could throw, and we're not handling that. The typical base exception
+     * types to this class do not give that guarantee either (.e.g, std::runtime_error)
+     * so the code does not bother.
+     *
+     * @tparam Args Format argument types
+     * @param [in] fmt Format string
+     * @param [in] args Format arguments
+     *
+     * @return std::string Format result
+     */
+    template <typename... Args>
+    static std::string failsafe_format(fmt::format_string<Args...> fmt, Args&&... args)
+    try
+    {
+        return fmt::format(fmt, std::forward<Args>(args)...);
+    }
+    catch (const std::exception& e)
+    {
+        std::string msg{"[Error while formatting the exception string]"};
+        msg += "\nFormat string: `";
+        msg += fmt.get().data();
+        msg += "`\nFormat error: `";
+        msg += e.what();
+        msg += '`';
+        return msg;
```

Comment:
why not use fmt::format to generate this message? would that cause problems?

Reply:
Using `fmt::format` in a `fmt::format` exception handler might cause trouble because we can't be sure that `fmt::format` is throwing because of what. It's a std::terminate() if it throws again, so I wanted to play safe. Ideally, this part of the code shouldn't construct anything that can throw, but that'd mean overengineering for this part. I think using std::string strikes a fine balance.

Reply:
alright, makes sense